crates/transpiler/src/passes/constrained_reschedule.rs

```diff
@@ -0,0 +1,304 @@
+            .iter()
+            .map(|c| c.index())
+            .collect();
+        (new_t1c, this_clbits)
+    } else {
+        (None, HashSet::new())
+    };
+    // Check immediate successors for overlap
+    for next_node_index in get_next_gate(dag, node_index) {
+        // Get the next node
+        let next_node = match dag.dag().node_weight(next_node_index) {
+            Some(NodeType::Operation(op)) => op,
+            _ => continue, // Skip non-operation nodes
+        };
+
+        // Compute next node start time separately for qreg and creg
+        let next_t0q_obj = node_start_time.get_item(next_node_index.index())?;
+        let next_t0q: f64 = next_t0q_obj
+            .as_ref()
+            .expect("Expected value in node_start_time for next_node_index")
+            .extract()?;
+
+        let next_qubits: HashSet<_> = dag
+            .qargs_interner()
+            .get(next_node.qubits)
+            .iter()
+            .map(|q| q.index())
+            .collect();
+
+        let next_op_view = next_node.op.view();
+        let (next_t0c, next_clbits) = if matches!(
+            next_op_view,
+            OperationRef::StandardInstruction(StandardInstruction::Measure)
+                | OperationRef::StandardInstruction(StandardInstruction::Reset)
+        ) {
+            // creg access starts after write latency
+            let next_t0c = Some(next_t0q + clbit_write_latency as f64);
+            let next_clbits: HashSet<_> = dag
+                .cargs_interner()
+                .get(next_node.clbits)
+                .iter()
+                .map(|c| c.index())
+                .collect();
+            (next_t0c, next_clbits)
+        } else {
+            (None, HashSet::new())
+        };
+
+        // Compute overlap if there is qubits overlap
+        let qreg_overlap = if !this_qubits.is_disjoint(&next_qubits) {
+            (new_t1q - next_t0q).max(0.0)
+        } else {
+            0.0
+        };
+
+        // Compute overlap if there is clbits overlap
+        let creg_overlap = if !this_clbits.is_empty()
+            && !next_clbits.is_empty()
+            && !this_clbits.is_disjoint(&next_clbits)
+        {
+            if let (Some(t1c), Some(t0c)) = (new_t1c, next_t0c) {
+                (t1c - t0c).max(0.0)
+            } else {
+                0.0
+            }
+        } else {
+            0.0
+        };
+
+        // Shift next node if there is finite overlap in either qubits or clbits
+        let overlap = qreg_overlap.max(creg_overlap);
+        if overlap > 0.0 {
+            let new_start_time = next_t0q + overlap;
+            node_start_time.set_item(next_node_index.index(), new_start_time)?;
+        }
+    }
+    Ok(())
+}
+
+#[pyfunction]
+#[pyo3(name="constrained_reschedule", signature=(dag, node_start_time, clbit_write_latency, acquire_align, pulse_align, target))]
+pub fn run_constrained_reschedule(
+    dag: &DAGCircuit,
+    node_start_time: &Bound<PyDict>,
```

Comment:
Is there a rust native type we can use for this? I realize this is coming from the property set in Python. But you're porting the scheduling passes to Rust right? I would expect the asap and alap scheduling passes to be returning a rust struct that is a pyclass that implements the mapping protocol so it looks like a dict to python, but internally it's a hashmap or something for rust so we can avoid the conversion and python over when accessing the start time data.